ompi/tools/ompi_info/ompi_info.1
.\" Man page contributed by Dirk Eddelbuettel <edd@debian.org>
.\" and released under the BSD license
.TH OMPI_INFO 1 "July 2006" "Open MPI" "OPEN MPI COMMANDS"
.SH NAME
ompi_info - Display information about the Open MPI installation
.
.\" **************************
.\"    Synopsis Section
.\" **************************
.SH SYNOPSIS
.B ompi_info [options]
.
.\" **************************
.\"    Description Section
.\" **************************
.SH DESCRIPTION
.PP
.B ompi_info
provides detailed information about the Open MPI installation.  It can
be useful for at least three common scenarios:
.PP
1. Checking local configuration and seeing how Open MPI was installed.
.PP
2. Submitting bug reports / help requests to the Open MPI community
(see 
.IR http://www.open-mpi.org/community/help/ )
.PP
3. Seeing a list of installed Open MPI plugins and querying what 
MCA parameters they support.
.
.\" **************************
.\"    Options Section
.\" **************************
.SH OPTIONS
.B ompi_info
accepts the following options:
.TP 8
.I \-a|\-\-all 
Show all configuration options and MCA parameters
.TP 8
.I \-\-arch   
Show architecture Open MPI was compiled on
.TP 8
.I \-c|\-\-config
Show configuration options
.TP 8
.I \-gmca|\-\-gmca <param> <value>
Pass global MCA parameters that are applicable to all contexts.
.TP 8
.I \-h|\-\-help
Shows help / usage message
.TP 8
.I \-\-hostname
Show the hostname that Open MPI was configured and built on
.TP 8
.I \-\-internal
Show internal MCA parameters (not meant to be modified by users)
.TP 8
.I \-mca|\-\-mca <param> <value>
Pass context-specific MCA parameters; they are considered global if --gmca is
not used and only one context is specified.
.TP 8
.I \-\-param <type> <component>
Show MCA parameters.  The first parameter is the type of the component
to display; the second parameter is the specific component to display
(or the keyword "all", meaning "display all components of this type").
.TP 8
.I \-\-parsable
When used in conjunction with other parameters, the output is
displayed in a machine-parsable format
.I \-\-parseable
Synonym for --parsable
.TP 8
.I \-\-path <type>
Show paths that Open MPI was configured with.  Accepts the following
parameters: prefix, bindir, libdir, incdir, pkglibdir, sysconfdir.
.TP 8
.I \-\-pretty
When used in conjunction with other parameters, the output is
displayed in 'prettyprint' format (default)
.TP 8
.I \-v|\-\-version <component> <scope>
Show version of Open MPI or a component.  <component> can be the
keywords "ompi" or "all", the name of a framework (e.g., "coll" shows
all components in the coll framework), or the name of a specific
component (e.g., "pls:rsh" shows the information from the rsh PLS
component).  <scope> can be one of: full, major, minor, release,
greek, svn.
.
.\" **************************
.\"    Examples Section
.\" **************************
.SH EXAMPLES
.TP 4
ompi_info
Show the default output of options and listing of installed
components in a human-readable / prettyprint format.
.
.
.TP
ompi_info --parsable
Show the default output of options and listing of installed
components in a machine-parsable format.
.
.
.TP
ompi_info --param btl openib
Show the MCA parameters of the "openib" BTL component in a
human-readable / prettyprint format.
.
.
.TP
ompi_info --param btl openib --parsable
Show the MCA parameters of the "openib" BTL component in a
machine-parsable format.
.
.
.TP
ompi_info --path bindir
Show the "bindir" that Open MPI was configured with.
.
.
.TP
ompi_info --version ompi full --parsable
Show the full version numbers of Open MPI (including the ORTE and OPAL
version numbers) in a machine-readable format.
.
.
.TP
ompi_info --version btl major
Show the major version number of all BTL components in a prettyprint
format.
.
.
.TP
ompi_info --version btl:tcp minor
Show the minor version number of the TCP BTL component in a
prettyprint format.
.
.
.TP
ompi_info --all
Show 
.I all 
information about the Open MPI installation, including all components
that can be found, the MCA parameters that they support, versions of
Open MPI and the components, etc.
.
.
.\" **************************
.\"    Authors Section
.\" **************************
.SH AUTHORS
The Open MPI maintainers -- see 
.I http://www.openmpi.org/
or the file
.IR AUTHORS .
.PP
This manual page was originally contributed by Dirk Eddelbuettel
<edd@debian.org>, one of the Debian GNU/Linux maintainers for Open
MPI, and may be used by others.


NEWS
Copyright (c) 2004-2007 The Trustees of Indiana University and Indiana
                        University Research and Technology
                        Corporation.  All rights reserved.
Copyright (c) 2004-2006 The University of Tennessee and The University
                        of Tennessee Research Foundation.  All rights
                        reserved.
Copyright (c) 2004-2007 High Performance Computing Center Stuttgart, 
                        University of Stuttgart.  All rights reserved.
Copyright (c) 2004-2006 The Regents of the University of California.
                        All rights reserved.
Copyright (c) 2006-2007 Cisco Systems, Inc.  All rights reserved.
Copyright (c) 2006      Voltaire, Inc. All rights reserved.
Copyright (c) 2006      Sun Microsystems, Inc.  All rights reserved.
                        Use is subject to license terms.
Copyright (c) 2006-2007 Los Alamos National Security, LLC.  All rights
                        reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of Open MPI since
version 1.0.

1.2.4
-----

- Nothing new yet.


1.2.3
-----

- Fix a regression in comm_spawn functionality that inadvertently
  caused the mapping of child processes to always start at the same
  place.  Thanks to Prakash Velayutham for helping discover the
  problem.
- Fix segfault when a user's home directory is unavailable on a remote
  node.  Thanks to Guillaume Thomas-Collignon for bringing the issue
  to our attention.
- Fix MPI_IPROBE to properly handle MPI_STATUS_IGNORE on mx and psm
  MTLs. Thanks to Sophia Corwell for finding this and supplying a
  reproducer.
- Fix some error messages in the tcp BTL.
- Use _NSGetEnviron instead of environ on Mac OS X so that there
  are no undefined symbols in the shared libraries.
- On OS X, when MACOSX_DEPLOYMENT_TARGET is 10.3 or higher, support
  building the Fortran 90 bindings as a shared library.  Thanks to
  Jack Howarth for his advice on making this work.
- No longer require extra include flag for the C++ bindings.
- Fix detection of weak symbols support with Intel compilers.
- Fix issue found by Josh England: ompi_info would not show framework
  MCA parameters set in the environment properly.
- Rename the oob_tcp_include/exclude MCA params to oob_tcp_if_include/exclude
  so that they match the naming convention of the btl_tcp_if_include/exclude
  params.  The old names are depreciated, but will still work.
- Add -wd as a synonym for the -wdir orterun/mpirun option.
- Fix the mvapi BTL to compile properly with compilers that do not support
  anonymous unions.  Thanks to Luis Kornblueh for reporting the bug.


1.2.2
-----

- Fix regression in 1.2.1 regarding the handling of $CC with both
  absolute and relative path names.
- Fix F90 array of status dimensions.  Thanks to Randy Bramley for
  noticing the problem.
- Add btl_openib_ib_pkey_value MCA parameter for controlling IB port selection.
- Fixed a variety of threading/locking bugs.
- Fixed some compiler warnings associated with ROMIO, OS X, and gridengine.
- If pbs-config can be found, use it to look for TM support.  Thanks
  to Bas van der Vlies for the inspiration and preliminary work.
- Fixed a deadlock in orterun when the rsh PLS encounters some errors.


1.2.1
-----

- Fixed a number of connection establishment errors in the TCP out-
  of-band messaging system.
- Fixed a memory leak when using mpi_comm calls.
  Thanks to Bas van der Vlies for reporting the problem.
- Fixed various memory leaks in OPAL and ORTE.
- Improved launch times when using TM (PBS Pro, Torque, Open PBS).
- Fixed mpi_leave_pinned to work for all datatypes.
- Fix functionality allowing users to disable sbrk() (the
  mpool_base_disable_sbrk MCA parameter) on platforms that support it.
- Fixed a pair of problems with the TCP "listen_thread" mode for the
  oob_tcp_listen_mode MCA parameter that would cause failures when
  attempting to launch applications.
- Fixed a segfault if there was a failure opening a BTL MX endpoint.
- Fixed a problem with mpirun's --nolocal option introduced in 1.2.
- Re-enabled MPI_COMM_SPAWN_MULTIPLE from singletons.
- LoadLeveler and TM configure fixes, Thanks to Martin Audet for the
  bug report.
- Various C++ MPI attributes fixes.
- Fixed issues with backtrace code on 64 bit Intel & PPC OS X builds.
- Fixed issues with multi-word CC variables and libtool.
  Thanks to Bert Wesarg for the bug reports.
- Fix issue with non-uniform node naming schemes in SLURM.
- Fix file descriptor leak in the Grid Engine/N1GE support.
- Fix compile error on OS X 10.3.x introduced with Open MPI 1.1.5.
- Implement MPI_TYPE_CREATE_DARRAY function (was in 1.1.5 but not 1.2).
- Recognize zsh shell when using rsh/ssh for launching MPI jobs.
- Ability to set the OPAL_DESTDIR or OPAL_PREFIX environment
  variables to "re-root" an existing Open MPI installation.
- Always include -I for Fortran compiles, even if the prefix is
  /usr/local.
- Support for "fork()" in MPI applications that use the
  OpenFabrics stack (OFED v1.2 or later).
- Support for setting specific limits on registered memory.


1.2
---

- Fixed race condition in the shared memory fifo's, which led to
  orphaned messages.
- Corrected the size of the shared memory file - subtracted out the
  space the header was occupying.
- Add support for MPI_2COMPLEX and MPI_2DOUBLE_COMPLEX.
- Always ensure to create $(includedir)/openmpi, even if the C++
  bindings are disabled so that the wrapper compilers don't point to
  a directory that doesn't exist.  Thanks to Martin Audet for
  identifying the problem.
- Fixes for endian handling in MPI process startup.
- Openib BTL initialization fixes for cases where MPI processes in the
  same job has different numbers of active ports on the same physical
  fabric.

ompi/tools/ompi_info/ompi_info.cc
//
// Copyright (c) 2004-2007 The Trustees of Indiana University and Indiana
//                         University Research and Technology
//                         Corporation.  All rights reserved.
// Copyright (c) 2004-2006 The University of Tennessee and The University
//                         of Tennessee Research Foundation.  All rights
//                         reserved.
// Copyright (c) 2004-2007 High Performance Computing Center Stuttgart, 
//                         University of Stuttgart.  All rights reserved.
// Copyright (c) 2004-2005 The Regents of the University of California.
//                         All rights reserved.
// Copryight (c) 2007      Cisco Systems, Inc.  All rights reserved.
// $COPYRIGHT$
// 
// Additional copyrights may follow
// 
// $HEADER$
//

#include "ompi_config.h"

#include <iostream>
#include <string>
#include <utility>
#include <list>

#include <stdio.h>
#include <ctype.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_NETDB_H
#include <netdb.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#include <errno.h>

#include "opal/mca/installdirs/installdirs.h"
#include "opal/class/opal_object.h"
#include "opal/runtime/opal.h"
#include "orte/runtime/runtime.h"
#include "opal/util/output.h"
#include "opal/util/cmd_line.h"
#include "opal/util/argv.h"
#include "opal/util/show_help.h"
#include "ompi/communicator/communicator.h"
#include "opal/mca/base/base.h"
#include "ompi/tools/ompi_info/ompi_info.h"
#include "opal/threads/mutex.h"

using namespace std;
using namespace ompi_info;


//
// Public variables
//

bool ompi_info::pretty = true;
opal_cmd_line_t *ompi_info::cmd_line = NULL;

const string ompi_info::type_all = "all";
const string ompi_info::type_ompi = "ompi";
const string ompi_info::type_orte = "orte";
const string ompi_info::type_opal = "opal";
const string ompi_info::type_base = "base";
ompi_info::type_vector_t ompi_info::mca_types;


int main(int argc, char *argv[])
{
  int ret = 0;
  bool want_help = false;
  bool cmd_error = false;
  bool acted = false;
  bool want_all = false;
  char **app_env = NULL, **global_env = NULL;
  int i, len;

  // Initialize the argv parsing handle
  if (OMPI_SUCCESS != opal_init_util()) {
    opal_show_help("help-ompi_info.txt", "lib-call-fail", true, 
                   "opal_init_util", __FILE__, __LINE__, NULL);
    exit(ret);
  }

  cmd_line = OBJ_NEW(opal_cmd_line_t);
  if (NULL == cmd_line) {
    ret = errno;
    opal_show_help("help-ompi_info.txt", "lib-call-fail", true, 
                   "opal_cmd_line_create", __FILE__, __LINE__, NULL);
    exit(ret);
  }

  opal_cmd_line_make_opt3(cmd_line, 'v', NULL, "version", 2, 
                          "Show version of Open MPI or a component.  The first parameter can be the keywords \"ompi\" or \"all\", a framework name (indicating all components in a framework), or a framework:component string (indicating a specific component).  The second parameter can be one of: full, major, minor, release, greek, svn.");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "param", 2, 
                          "Show MCA parameters.  The first parameter is the framework (or the keyword \"all\"); the second parameter is the specific component name (or the keyword \"all\").");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "internal", 0, 
                          "Show internal MCA parameters (not meant to be modified by users)");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "path", 1, 
                          "Show paths that Open MPI was configured with.  Accepts the following parameters: prefix, bindir, libdir, incdir, pkglibdir, sysconfdir");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "arch", 0, 
                          "Show architecture Open MPI was compiled on");
  opal_cmd_line_make_opt3(cmd_line, 'c', NULL, "config", 0, 
                          "Show configuration options");
  opal_cmd_line_make_opt3(cmd_line, 'h', NULL, "help", 0, 
                          "Show this help message");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "pretty", 0, 
                          "When used in conjunction with other parameters, the output is displayed in 'prettyprint' format (default)");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "parsable", 0, 
                          "When used in conjunction with other parameters, the output is displayed in a machine-parsable format");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "parseable", 0, 
                          "Synonym for --parsable");
  opal_cmd_line_make_opt3(cmd_line, '\0', NULL, "hostname", 0, 
                          "Show the hostname that Open MPI was configured "
                          "and built on");
  opal_cmd_line_make_opt3(cmd_line, 'a', NULL, "all", 0, 
                          "Show all configuration options and MCA parameters");

  // Call some useless functions in order to guarantee to link in some
  // global variables.  Only check the return value so that the
  // compiler doesn't optimize out the useless function.

  if (OMPI_SUCCESS != ompi_comm_link_function()) {
    // Stop .. or I'll say stop again!
    ++ret;
  } else {
    --ret;
  }

  // set our threading level
  opal_set_using_threads(false);

  // Get MCA parameters, if any */
  
  if( OMPI_SUCCESS != mca_base_open() ) {
      opal_show_help("help-ompi_info.txt", "lib-call-fail", true, "mca_base_open", __FILE__, __LINE__ );
      exit(1);
  }
  mca_base_cmd_line_setup(cmd_line);

  // Do the parsing

  if (OMPI_SUCCESS != opal_cmd_line_parse(cmd_line, false, argc, argv)) {
      cmd_error = true;
  }
  if (!cmd_error && 
        (opal_cmd_line_is_taken(cmd_line, "help") || 
         opal_cmd_line_is_taken(cmd_line, "h"))) {
      want_help = true;
  }
  if (cmd_error || want_help) {
      char *usage = opal_cmd_line_get_usage_msg(cmd_line);
      opal_show_help("help-ompi_info.txt", "usage", true, usage);
      free(usage);
      exit(cmd_error ? 1 : 0);
  }

  mca_base_cmd_line_process_args(cmd_line, &app_env, &global_env);

  // putenv() all the stuff that we got back from env (in case the
  // user specified some --mca params on the command line).  This
  // creates a memory leak, but that's unfortunately how putenv()
  // works.  :-(

  len = opal_argv_count(app_env);
  for (i = 0; i < len; ++i) {
      putenv(app_env[i]);
  }
  len = opal_argv_count(global_env);
  for (i = 0; i < len; ++i) {
      putenv(global_env[i]);
  }

  ompi_info::mca_types.push_back("mca");
  ompi_info::mca_types.push_back("mpi");
  ompi_info::mca_types.push_back("orte");
  ompi_info::mca_types.push_back("opal");

  ompi_info::mca_types.push_back("backtrace");
  ompi_info::mca_types.push_back("memory");
  ompi_info::mca_types.push_back("paffinity");
  ompi_info::mca_types.push_back("maffinity");
  ompi_info::mca_types.push_back("timer");
  ompi_info::mca_types.push_back("installdirs");
#if OPAL_ENABLE_FT == 1
  ompi_info::mca_types.push_back("crs");
#endif

  ompi_info::mca_types.push_back("allocator");
  ompi_info::mca_types.push_back("coll");
  ompi_info::mca_types.push_back("io");
  ompi_info::mca_types.push_back("mpool");
  ompi_info::mca_types.push_back("pml");
  ompi_info::mca_types.push_back("bml");
  ompi_info::mca_types.push_back("rcache");
  ompi_info::mca_types.push_back("btl");
  ompi_info::mca_types.push_back("mtl");
  ompi_info::mca_types.push_back("topo");
  ompi_info::mca_types.push_back("osc");
  ompi_info::mca_types.push_back("common");
#if OPAL_ENABLE_FT == 1
  ompi_info::mca_types.push_back("crcp");
#endif

  ompi_info::mca_types.push_back("errmgr");
  ompi_info::mca_types.push_back("gpr");
  ompi_info::mca_types.push_back("iof");
  ompi_info::mca_types.push_back("ns");
  ompi_info::mca_types.push_back("oob");
  ompi_info::mca_types.push_back("odls");
  ompi_info::mca_types.push_back("ras");
  ompi_info::mca_types.push_back("rds");
  ompi_info::mca_types.push_back("rmaps");
  ompi_info::mca_types.push_back("rmgr");
  ompi_info::mca_types.push_back("rml");
  ompi_info::mca_types.push_back("pls");
  ompi_info::mca_types.push_back("sds");
  ompi_info::mca_types.push_back("soh");
#if OPAL_ENABLE_FT == 1
  ompi_info::mca_types.push_back("snapc");
#endif
  ompi_info::mca_types.push_back("filem");

  // Execute the desired action(s)

  if (opal_cmd_line_is_taken(cmd_line, "pretty")) {
    ompi_info::pretty = true;
  } else if (opal_cmd_line_is_taken(cmd_line, "parsable") || opal_cmd_line_is_taken(cmd_line, "parseable")) {
    ompi_info::pretty = false;
  }

  want_all = opal_cmd_line_is_taken(cmd_line, "all");
  if (want_all || opal_cmd_line_is_taken(cmd_line, "version")) {
    do_version(want_all, cmd_line);
    acted = true;
  }
  if (want_all || opal_cmd_line_is_taken(cmd_line, "path")) {
    do_path(want_all, cmd_line);
    acted = true;
  }
  if (want_all || opal_cmd_line_is_taken(cmd_line, "arch")) {
    do_arch();
    acted = true;
  }
  if (want_all || opal_cmd_line_is_taken(cmd_line, "hostname")) {
    do_hostname();
    acted = true;
  }
  if (want_all || opal_cmd_line_is_taken(cmd_line, "config")) {
    do_config(true);
    acted = true;
  }
  if (want_all || opal_cmd_line_is_taken(cmd_line, "param")) {
    do_params(want_all, opal_cmd_line_is_taken(cmd_line, "internal"));
    acted = true;
  }

  // If no command line args are specified, show default set

  if (!acted) {
    ompi_info::show_ompi_version(ver_full);
    ompi_info::show_path(path_prefix, opal_install_dirs.prefix);
    ompi_info::do_arch();
    ompi_info::do_hostname();
    ompi_info::do_config(false);
    ompi_info::open_components();
    for (ompi_info::type_vector_t::size_type i = 0; 
         i < mca_types.size(); ++i) {
        if ("mpi" != mca_types[i]) {
            ompi_info::show_component_version(mca_types[i], component_all, 
                                              ver_full, type_all);
        }
    }
  }

  // All done

  if (NULL != app_env) {
      opal_argv_free(app_env);
  }
  if (NULL != global_env) {
      opal_argv_free(global_env);
  }
  ompi_info::close_components();
  OBJ_RELEASE(cmd_line);
  mca_base_close();

  opal_finalize_util();

  return 0;
}
